src/surface.cpp

```diff
@@ -983,6 +1004,212 @@ void SurfaceQuadric::to_hdf5_inner(hid_t group_id) const
   write_dataset(group_id, "coefficients", coeffs);
 }
 
+//==============================================================================
+// Torus helper functions
+//==============================================================================
+
+double torus_distance(double x1, double x2, double x3, double u1, double u2,
+  double u3, double A, double B, double C, bool coincident)
+{
+  // Coefficients for equation: (c2 t^2 + c1 t + c0)^2 = c2' t^2 + c1' t + c0'
+  double D = (C * C) / (B * B);
+  double c2 = u1 * u1 + u2 * u2 + D * u3 * u3;
+  double c1 = 2 * (u1 * x1 + u2 * x2 + D * u3 * x3);
+  double c0 = x1 * x1 + x2 * x2 + D * x3 * x3 + A * A - C * C;
+  double four_A2 = 4 * A * A;
+  double c2p = four_A2 * (u1 * u1 + u2 * u2);
+  double c1p = 2 * four_A2 * (u1 * x1 + u2 * x2);
+  double c0p = four_A2 * (x1 * x1 + x2 * x2);
+
+  // Coefficient for equation: a t^4 + b t^3 + c t^2 + d t + e = 0
+  double coeff[5];
+  coeff[0] = c0 * c0 - c0p;
+  coeff[1] = 2 * c0 * c1 - c1p;
+  coeff[2] = c1 * c1 + 2 * c0 * c2 - c2p;
+  coeff[3] = 2 * c1 * c2;
+  coeff[4] = c2 * c2;
+
+  std::complex<double> roots[4];
+  oqs::quartic_solver(coeff, roots);
+
+  // Find smallest positive, real root. In the case where the particle is
+  // coincident with the surface, we are sure to have one root very close to
+  // zero but possibly small and positive. A tolerance is set to discard that
+  // zero.
+  double distance = INFTY;
+  double cutoff = coincident ? 1e-9 : 0.0;
```

Comment:
Would one of the already defined constants (i.e. `FP_COINCIDENT`) work here or is `1e-9` a sweet spot for this check?

Reply:
Unfortunately not. I was running into cases where I'd get a root greater than `FP_COINCIDENT` when the particle was coincident. A lower cutoff would probably still work (e.g., 1e-10), but I might have a better solution anyway in the coincident case, which is to just solve a cubic instead since we know one root should be zero.

Reply:
An update here: I did try out the approach of solving a cubic equation when the particle is coincident with the surface, but it didn't seem to yield any measurable increase in performance. Given that it would require adding more code, I'm just going to stick with our current approach, which is to solve the quartic anyway and skip the expected root near zero.

As far as the value of the cutoff, I ran the regression test model with a few million neutrons and the highest value that I got for the expected zero root was about 1.3e-11, so it's probably safe to reduce the cutoff here to 1e-10. Let me know how you'd like me to proceed.

Reply:
> An update here: I did try out the approach of solving a cubic equation when the particle is coincident with the surface, but it didn't seem to yield any measurable increase in performance. Given that it would require adding more code, I'm just going to stick with our current approach, which is to solve the quartic anyway and skip the expected root near zero.

Thanks for testing that out! 

> As far as the value of the cutoff, I ran the regression test model with a few million neutrons and the highest value that I got for the expected zero root was about 1.3e-11, so it's probably safe to reduce the cutoff here to 1e-10. Let me know how you'd like me to proceed.

I'd say yeah let's reduce it to `1e-10` then. Thanks for pinning that down. Normally 
I'd request a new constant for clarity, but your comment right above is pretty thorough.